Mage.Sets/src/mage/cards/k/KnickknackOuphe.java

```diff
@@ -0,0 +1,122 @@
+        this.addAbility(new EntersBattlefieldAbility(new EntersBattlefieldWithXCountersEffect(CounterType.P1P1.createInstance())));
+
+        // When Knickknack Ouphe enters the battlefield, reveal the top X cards of your library. You may put any number of Aura cards with mana value X or less from among them onto the battlefield. Then put all cards revealed this way that weren't put onto the battlefield on the bottom of your library in a random order.
+        this.addAbility(new EntersBattlefieldAbility(new KnickknackOuphePutOntoBattlefieldEffect()));
+
+    }
+
+    private KnickknackOuphe(final KnickknackOuphe card) {
+        super(card);
+    }
+
+    @Override
+    public KnickknackOuphe copy() {
+        return new KnickknackOuphe(this);
+    }
+}
+
+class KnickknackOuphePutOntoBattlefieldEffect extends OneShotEffect {
+
+    public KnickknackOuphePutOntoBattlefieldEffect() {
+        super(Outcome.PutCardInPlay);
+        staticText = "reveal the top X cards of your library. " + 
+        "You may put any number of Aura cards with mana value X or less from among them onto the battlefield. " + 
+        "Then put all cards revealed this way that weren't put onto the battlefield on the bottom of your library in a random order";
+    }
+
+    public KnickknackOuphePutOntoBattlefieldEffect(final KnickknackOuphePutOntoBattlefieldEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller != null) {
+            int count = ManacostVariableValue.ETB.calculate(game, source, null);
+            if (count > 0) {
+                Cards cards = new CardsImpl(controller.getLibrary().getTopCards(game, count));
+                controller.revealCards(source, cards, game);
+
+                FilterCard filter = new FilterEnchantmentCard("Aura cards with mana value " + count + " or less to put onto the battlefield");
+                filter.add(SubType.AURA.getPredicate());
+                filter.add(new ManaValuePredicate(ComparisonType.OR_LESS, count));
+                
+                if (cards.count(filter, controller.getId(), source, game) > 0) {
+                    TargetCard targetAuras = new TargetCard(0, count, Zone.LIBRARY, filter);
+                    targetAuras.setRequired(false);
+
+                    if (controller.choose(Outcome.PutCardInPlay, cards, targetAuras, source, game)) {
+                        targetAuras.getTargets().stream().forEach(t -> {
+                            Card card = cards.get(t, game);
+                            if (card != null) {
+                                cards.remove(card);
+                                controller.moveCards(card, Zone.BATTLEFIELD, source, game);
+                            }
+                        });
+
+                        targetAuras.clearChosen();
```

Comment:
just curious what this line is for?

Reply:
Not sure, copied the implementation from Genesis Hydra, so if it's unnecessary here than I apologize